velox/functions/prestosql/types/TimestampWithTimeZoneType.cpp

```diff
@@ -14,8 +14,157 @@
+#include "velox/type/tz/TimeZoneMap.h"
 
 namespace facebook::velox {
+namespace {
+void castToTsWithTZFromTimestamp(
+    const SimpleVector<Timestamp>* inputVector,
+    exec::EvalCtx& context,
+    const SelectivityVector& rows,
+    RowVector* rowResult) {
+  const auto& config = context.execCtx()->queryCtx()->queryConfig();
+  const auto sessionTzName = config.sessionTimezone();
+  int64_t sessionTzID = 0;
+  if (!sessionTzName.empty()) {
+    sessionTzID = util::getTimeZoneID(sessionTzName);
+  }
+
+  auto timestampVector = rowResult->childAt(0)->asFlatVector<int64_t>();
+  auto timezoneVector = rowResult->childAt(1)->asFlatVector<int16_t>();
+  auto rawTsValues = timestampVector->values()->asMutable<int64_t>();
+  auto rawTzValues = timezoneVector->values()->asMutable<int16_t>();
+
+  context.applyToSelectedNoThrow(rows, [&](auto row) {
+    if (inputVector->isNullAt(row)) {
+      rowResult->setNull(row, true);
+    } else {
+      auto ts = inputVector->valueAt(row);
+      if (!config.adjustTimestampToTimezone()) {
+        ts.toGMT(sessionTzID);
+      }
+      rawTsValues[row] = ts.toMillis();
+      rawTzValues[row] = sessionTzID;
+    }
+  });
+}
+
+template <TypeKind kind>
+void castToTimestampWithTimeZone(
+    const BaseVector& input,
+    exec::EvalCtx& context,
+    const SelectivityVector& rows,
+    RowVector* rowResult) {
+  if constexpr (kind == TypeKind::TIMESTAMP) {
+    castToTsWithTZFromTimestamp(
+        input.as<SimpleVector<Timestamp>>(), context, rows, rowResult);
+  }
+}
+
+void castFromTsWithTZToTimestamp(
+    const RowVector* inputVector,
+    exec::EvalCtx& context,
+    const SelectivityVector& rows,
+    FlatVector<Timestamp>* flatResult) {
+  const auto& config = context.execCtx()->queryCtx()->queryConfig();
+  const auto timestampVector =
+      inputVector->childAt(0)->as<SimpleVector<int64_t>>();
+  const auto timezoneVector =
+      inputVector->childAt(1)->as<SimpleVector<int16_t>>();
+
+  context.applyToSelectedNoThrow(rows, [&](auto row) {
+    if (inputVector->isNullAt(row)) {
+      flatResult->setNull(row, true);
+    } else {
+      Timestamp ts = Timestamp::fromMillis(timestampVector->valueAt(row));
+      if (!config.adjustTimestampToTimezone()) {
+        ts.toTimezone(timezoneVector->valueAt(row));
+      }
+      flatResult->set(row, ts);
+    }
+  });
+}
+
+template <TypeKind kind>
+void castFromTimestampWithTimeZone(
+    const BaseVector& input,
+    exec::EvalCtx& context,
+    const SelectivityVector& rows,
+    VectorPtr& result) {
+  const auto inputVector = input.as<RowVector>();
+
+  if constexpr (kind == TypeKind::TIMESTAMP) {
+    castFromTsWithTZToTimestamp(
+        inputVector, context, rows, result->as<FlatVector<Timestamp>>());
+  }
+}
+} // namespace
+
+bool TimestampWithTimeZoneCastOperator::isSupportedFromType(
+    const TypePtr& other) const {
+  switch (other->kind()) {
+    case TypeKind::TIMESTAMP:
```

Comment:
Looks like only casts to and from TIMESTAMP are supported. Would you update PR description to make this clear?